x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,378 @@
+func (m *MetricSet) createRequest(dateStr, apiKey string) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, m.config.APIURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	q := req.URL.Query()
+	q.Add("date", dateStr)
+	req.URL.RawQuery = q.Encode()
+
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", apiKey))
+	for key, value := range m.headers {
+		req.Header.Add(key, value)
+	}
+
+	return req, nil
+}
+
+// processResponse handles the API response and processes the usage data.
+func (m *MetricSet) processResponse(resp *http.Response, dateStr string) error {
+	var usageResponse UsageResponse
+	if err := json.NewDecoder(resp.Body).Decode(&usageResponse); err != nil {
+		return fmt.Errorf("error decoding response: %w", err)
+	}
+
+	m.logger.Info("Fetched usage metrics for date:", dateStr)
+
+	events := make([]mb.Event, 0, len(usageResponse.Data))
+
+	m.processUsageData(events, usageResponse.Data)
+	m.processDalleData(events, usageResponse.DalleApiData)
+	m.processWhisperData(events, usageResponse.WhisperApiData)
+	m.processTTSData(events, usageResponse.TtsApiData)
+
+	// Process additional data.
+	//
+	// NOTE(shmsr): During testing, could not get the usage data for the following
+	// and found no documentation, example responses, etc. That's why let's store them
+	// as it is so that we can use processors later on to process them as needed.
+	m.processFTData(events, usageResponse.FtData)
+	m.processAssistantCodeInterpreterData(events, usageResponse.AssistantCodeInterpreterData)
+	m.processRetrievalStorageData(events, usageResponse.RetrievalStorageData)
+
+	return nil
+}
+
+func (m *MetricSet) processUsageData(events []mb.Event, data []UsageData) {
+
+	for _, usage := range data {
+		event := mb.Event{
+			MetricSetFields: mapstr.M{
+				"data": mapstr.M{
+					"organization_id":               usage.OrganizationID,
+					"organization_name":             usage.OrganizationName,
+					"aggregation_timestamp":         time.Unix(usage.AggregationTimestamp, 0).UTC(), // epoch time to time.Time (UTC)
+					"n_requests":                    usage.NRequests,
+					"operation":                     usage.Operation,
+					"snapshot_id":                   usage.SnapshotID,
+					"n_context_tokens_total":        usage.NContextTokensTotal,
+					"n_generated_tokens_total":      usage.NGeneratedTokensTotal,
+					"email":                         usage.Email,
+					"api_key_id":                    usage.ApiKeyID,
+					"api_key_name":                  usage.ApiKeyName,
+					"api_key_redacted":              usage.ApiKeyRedacted,
+					"api_key_type":                  usage.ApiKeyType,
+					"project_id":                    usage.ProjectID,
+					"project_name":                  usage.ProjectName,
+					"request_type":                  usage.RequestType,
+					"n_cached_context_tokens_total": usage.NCachedContextTokensTotal,
+				},
```

Comment:
Does this API provide a division of Input and Output tokens?
If yes, we can include that also.

Reply:
I didn't get this. Can you please elaborate?

Reply:
The tokens received are actually input (prompt related) and output (response related) token. 
See this for example. 
Do we have such information [here](https://www.elastic.co/docs/current/integrations/gcp_vertexai#metrics-reference) for the tokens ?
That information helps us better estimate if we can improve the prompt or response.

Reply:
Oh `"division" of Input and Output tokens` confused me. Yeah it is already there otherwise they would have been no point in making this module if the most basic stats of the usage we cannot get.

See these fields: `n_context_tokens_total` and `n_generated_tokens_total` in this very structure where the former is input and latter is output tokens.

For an example query like this:
```json
"model": "gpt-4o-mini-2024-07-18",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Tell me about Elastic in one sentence."
      }
    ]
```

with a response like this:

```json
{
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "Elastic is a company known for its proprietary software solutions, including Elasticsearch, that enable powerful search, logging, and data analytics capabilities across a range of applications and industries.",
        "refusal": null
      },
      "logprobs": null,
      "finish_reason": "stop"
    }
```

Here are the values:
- `n_context_tokens_total`: 25
- `n_generated_tokens_total`: 33


See more examples values here: https://github.com/elastic/observability-dev/issues/3918#issuecomment-2333385190

Reply:
`n_cached_context_tokens_total` one you pointed is likely empty always. I've never seen it fill even once so far.